saiph/reduction/utils/svd.py

```diff
@@ -1,26 +1,121 @@
-from typing import Any, Tuple
+from typing import Any, Optional, Tuple
 
+import numpy as np
 import pandas as pd
 from numpy.typing import NDArray
 from scipy import linalg
 from sklearn.utils import extmath
 
 
-def SVD(
-    df: pd.DataFrame, svd_flip: bool = True
+def get_svd(
+    df: pd.DataFrame,
+    nf: Optional[int] = None,
+    *,
+    svd_flip: bool = True,
+    seed: Optional[int] = None,
 ) -> Tuple[NDArray[Any], NDArray[Any], NDArray[Any]]:
     """Compute Singular Value Decomposition.
 
-    Parameters:
-        df: Matrix to decompose.
-        svd_flip: Whether to use svd_flip on U and V or not.
+    Arguments
+    ---------
+        df: Matrix to decompose, shape (m, n)
+        nf: target number of dimensions to retain (number of singular values). Default `None`
+        svd_flip: Whether to use svd_flip on U and V or not. Default `True`
+        seed: random seed. Default `None`
 
-    Returns:
-        U: Unitary matrix having left singular vectors as columns.
-        s: The singular values.
-        V: Unitary matrix having right singular vectors as rows.
+    Returns
+    -------
+        U: unitary matrix having left singular vectors as columns, shape (m,l)
+        S: vector of the singular values, shape (l,)
+        Vt: unitary matrix having right singular vectors as rows, shape (l,n)
     """
-    U, s, V = linalg.svd(df, full_matrices=False)
+    if nf is not None and nf != np.min(df.shape):
```

Comment:
As discussed offline, there is a threshold in nf beyond which it is better to use the full svd. So this condition is probably not the best. Ideally, the threshold on nf should be backed by some evidence (experiments?).

Also, if we decide to have a way to force the use of either the full_svd or randomized_svd, we should only do this for the auto_svd (see previous comment)